### ISSUE TITLE
feat: incremental learning con memoria acotada para training

### ISSUE THREAD

**Issue by clarikearney** [fedehofmann/oil_and_gas_mlops_pipeline#36]
## Contexto

`RandomForestRegressor.fit()` y `get_historical_features` de Feast cargan el dataset completo en RAM al entrenar. Esto **limita el rango de fechas con el que se puede entrenar localmente** a aproximadamente 1 año (`date_from=2023-01-01`, `date_to=2023-12-31`). Para histórico completo (~2019 a hoy) se rompe por OOM.

Este issue es la continuación natural del trabajo de #31 (Evidently AI / model decay), donde quedó documentada la limitación en la bitácora del roadmap.

### El problema con Random Forest no es de implementación, es algorítmico

Random Forest es un **bagging ensemble**: arma N árboles independientes, cada uno entrenado con un bootstrap sample (muestreo aleatorio con reemplazo) del **dataset completo**. Para tomar esos samples necesita acceso simultáneo a toda la data al armar cada árbol. Por eso `RandomForestRegressor` no tiene `partial_fit` y **no puede tenerlo**: la limitación está en cómo funciona el algoritmo, no en cómo está implementado.

```
RF actual:
  fit(dataset_completo)  ← carga TODO en RAM, una sola vez
                            (~5 GB de dataset + 50 MB de modelo final)
```

Si el dataset > RAM → OOM, sin escape posible.

### Por qué XGBoost sí lo soporta

XGBoost es un **gradient boosting ensemble** construido secuencialmente: cada árbol nuevo aprende del **error residual** del modelo anterior. Esa dependencia secuencial es exactamente lo que permite hacer continuation:

```
Chunk de enero:    entreno 10 árboles → modelo M_enero
Chunk de febrero:  cargo M_enero, entreno 10 árboles más sobre el residuo → M_febrero
Chunk de marzo:    cargo M_febrero, agrego 10 árboles → M_marzo
...
```

Cada batch agrega árboles al final del modelo sin tocar los anteriores. La memoria queda acotada por el tamaño del chunk:

```
XGBoost incremental:
  for chunk in chunks_mensuales:
      fit(chunk, xgb_model=modelo_previo)  ← solo el chunk + modelo previo en RAM
                                              (~500 MB de chunk + ~10 MB de modelo)
```

### Cuenta de memoria comparada

| Esquema | Memoria peak | Limitante |
|---|---|---|
| RF + dataset completo (actual) | `sizeof(dataset) + sizeof(modelo) ≈ 5 GB + 50 MB` | Si el dataset > RAM → OOM, sin escape posible. |
| XGBoost incremental por chunks | `sizeof(chunk) + sizeof(modelo_creciente) ≈ 500 MB + ~10 MB` | Acotada por `chunk_size`, independiente del dataset total. |

### Aclaración importante: cambiar el modelo solo NO mejora memoria

La mejora viene de **dos cambios juntos**:

1. **Cambiar el algoritmo** → uno que soporte continuación (XGBoost vía `xgb_model` param, o SGDRegressor vía `partial_fit`).
2. **Cambiar el bucle de entrenamiento** → de "un fit con todo el dataset" a "N fits, uno por chunk".

Si se cambia el modelo a XGBoost pero se hace `xgb.fit(dataset_completo)`, se reproduce la misma OOM. El modelo es el habilitador; el bucle es el que reduce la memoria.

## Por qué este issue importa

- **Producción a escala:** un sistema de MLOps que solo soporta entrenar con un año de datos no es viable productivamente. Vaca Muerta tiene historia desde ~2012, y el régimen estable arranca en 2019. Dejar fuera 5+ años de datos es perder señal.
- **Convergencia a la media en inferencia futura** (limitación documentada en README) se atenuaría con más datos de entrenamiento.

## Lo que NO resuelve este issue

- El OOM en `prepare_offline_store` (carga del CSV completo + groupby + rolling). Eso es un cuello de botella separado, antes del training. **Issue separado a abrir**: refactorear `prepare_offline_store` a procesamiento por chunks o usar Dask/Polars en lugar de pandas in-memory.
- El OOM de la presión total de containers (Ray Serve + entrenamientos + Evidently). Ese se resuelve con más RAM en Docker Desktop o bajando `num_replicas` de Ray Serve.

### 1. Modelo: XGBoost (sobre SGDRegressor)

| Opción | Pros | Contras |
|---|---|---|
| **SGDRegressor** (sklearn `partial_fit`) | Memoria estrictamente constante. API simple. Concepto canónico. | **Lineal** — pierde no-linealidad de RF. Caída esperable de R². |
| **XGBoost** (con `xgb_model` param) ✅ | Mantiene capacidad no-lineal. Estándar industrial. | El modelo crece con cada batch (más árboles). Memoria por batch acotada (no estrictamente constante). |

**Decisión: XGBoost.** En este problema con features no-lineales (rolling means, profundidad, tipo de extracción), la capacidad no-lineal es más valiosa que la diferencia marginal de memoria entre los dos esquemas. En ambos casos la memoria queda acotada por chunk; el modelo XGBoost final pesa < 10 MB con `max_depth=6`.

### 2. Alcance: reemplazo total de RandomForest

Los 10 experimentos de la lista `EXPERIMENTS` pasan a usar XGBoost. La API de inferencia (`api/main.py`) cambia a `mlflow.xgboost.load_model`. Más simple operacionalmente que coexistencia (RF + XGBoost), evita branches en `select_best_model` y mantiene el repo coherente.

### 3. Versión y API

- `xgboost==3.2.0` (estable, sin conflictos con MLFlow 3.11 + scikit-learn 1.8 + Feast 0.47).
- **Sklearn-style API**: `XGBRegressor.fit(X, y, xgb_model=path_or_booster)`. Mantiene `feature_names_in_` y `predict()` con la misma interfaz que sklearn.
- **Logueo MLFlow**: `mlflow.xgboost.log_model(model_format="ubj")` en lugar de `mlflow.sklearn.log_model`. El formato nativo `.ubj` es portable entre versiones; pickle es frágil. Además preserva metadata categórica si en el futuro se migra a `enable_categorical=True`.
- **Carga MLFlow**: `mlflow.xgboost.load_model(uri)` en `api/main.py`.

### 4. Hiperparámetros

| RandomForest (actual) | XGBoost (target) | Notas |
|---|---|---|
| `n_estimators=100` | `n_estimators=10` por chunk | **En continuation, `n_estimators` es árboles a agregar, no total.** Con chunks mensuales (12 al año) y `n_estimators=10` por chunk, se llega a ~120 árboles totales. |
| `max_depth=10` o `None` | `max_depth=6` (default XGB) | Misma semántica. XGBoost tiene un default más conservador que tiende a generalizar mejor. |
| `random_state=42` | `random_state=42` | Sin cambio. |
| (no aplica) | `learning_rate=0.1` | RF no tiene equivalente. Default de XGB es 0.3 — agresivo para incremental. **0.1** es más estable: ningún chunk individual domina demasiado al modelo final. |

### 5. Chunks

- **Tamaño:** mensual (1 mes por chunk).
- **Orden:** temporal (cronológico), no shuffle.
- **Justificación:** alinea con el split temporal del DAG (train = 80% más antiguo, test = 20% más reciente). Los meses recientes pesan más en el modelo final, lo cual es coherente con el caso de uso de inferencia (predecir mes siguiente).

### 6. Categóricas: mantener `LabelEncoder` por ahora

`tipoextraccion` se sigue encodando con `LabelEncoder` (entero ordinal). XGBoost lo va a tratar como ordinal numérico — es subóptimo pero funciona. Migrar a `enable_categorical=True` con `pd.Categorical` de universo fijo queda para un PR separado, no acoplado a este refactor.

### 7. Issue #13 (LabelEncoder como artefacto MLFlow): separado

No se incluye en este PR. Este issue ya tiene mucho cambio; #13 puede ir después sin bloquear nada.

## Pitfalls mapeados antes de codear

1. **Olvidarse de `xgb_model` en `.fit()` reentrena desde cero silenciosamente** — sin warning. Hay que asertarlo o testearlo.
2. **`n_estimators` en continuation = árboles a agregar, no total**. Si dejo 100, después de 12 chunks tengo 1200 árboles, no 100.
3. **Schema de columnas inconsistente entre chunks**: si en un chunk las columnas vienen en otro orden, `xgb_model` falla. Asegurar `X = X[FEATURES_LIST]` antes de cada `.fit()`.
4. **Pickle vs `.ubj`**: si se usa `mlflow.sklearn.log_model` en lugar de `mlflow.xgboost.log_model`, se pierde metadata específica de XGBoost y se rompe portabilidad entre versiones.
5. **Drift temporal por orden de chunks**: el último chunk pesa más. Mitigado bajando `learning_rate` a 0.1.
6. **Mismatch de hiperparámetros estructurales** entre chunks (ej. cambiar `max_depth` mid-training) puede generar un modelo internamente inconsistente. Mantener constantes los structural params.

## Plan de implementación

1. Crear branch `feature/incremental-learning-xgboost` desde `main` actualizado (después de mergear #32, #33, #34, #35).
2. Refactor de `train_model`:
   - Iterar el dataset por chunks mensuales (vía `event_timestamp`).
   - Para cada chunk: `XGBRegressor(...).fit(X_chunk, y_chunk, xgb_model=prev_booster_path)`.
   - Guardar el booster en disco después de cada chunk (formato `.ubj`).
3. Configurar `learning_rate=0.1`, `n_estimators=10` por chunk, `max_depth=6`, `random_state=42`.
4. Migrar `evaluate_model`: cambiar `pickle.load(model_path)` a `xgb.XGBRegressor().load_model(model_path)` o equivalente.
5. Cambiar `mlflow.sklearn.log_model` → `mlflow.xgboost.log_model(model_format="ubj")` en `evaluate_model`.
6. Cambiar `mlflow.sklearn.load_model` → `mlflow.xgboost.load_model` en `api/main.py`.
7. Adaptar la lista `EXPERIMENTS` con los nuevos hiperparámetros (variar `n_estimators` por chunk: 5/10/15, `max_depth`: 4/6/8, features reducidas).
8. Actualizar `monitor_model`: verificar que sigue funcionando (Evidently no depende del tipo de modelo, solo de las predicciones, así que debería ser transparente).
9. Validar end-to-end con dataset de 2023 + smoke test con dataset más grande (ej. 2022-2023).
10. Documentar decisión de diseño #13 en README (por qué XGBoost, trade-off vs RF, cómo se entrena en chunks, hiperparámetros distintos).
11. Sumar entrada cerrada a la bitácora del roadmap (con resultados reales del experimento).

## Pendiente al cerrar

- Issue separado para refactorear `prepare_offline_store` con procesamiento por chunks (hoy carga el CSV completo).
- Opcional: bajar `num_replicas` en Ray Serve (cambia decisión #10 del README) para liberar RAM y no necesitar parar `api-1` durante el DAG.
- Opcional: migrar `LabelEncoder` → `enable_categorical=True` con `pd.Categorical` de universo fijo, para que XGBoost trate `tipoextraccion` como categórica nominal (no ordinal).

## Origen

Bitácora de [`roadmap.md`](../blob/main/roadmap.md), sección "En desarrollo — #31" → "Próximos features e issues a abrir".
